File: java-workshop/part-4-behavior/src/test/java/service/PaymentServiceTest.java
package service;

import domain.PaymentRequest;
import domain.PaymentResult;
import domain.PaymentStatus;
import gateway.PaymentGateway;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.verifyNoInteractions;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
class PaymentServiceTest {

    @Mock
    private PaymentGateway paymentGateway;

    private GatewayPaymentService paymentService;

    @BeforeEach
    void setUp() {
        paymentService = new GatewayPaymentService(paymentGateway);
    }

    @Test
    void processPayment_returnsSuccess_whenGatewayAcceptsPayment() {
        when(paymentGateway.process(any(PaymentRequest.class)))
            .thenReturn(new PaymentResult("ORDER-001", PaymentStatus.SUCCESS, "TXN-123"));

        var result = paymentService.processPayment("ORDER-001", "CUST-001", 29.0);

        assertEquals(PaymentStatus.SUCCESS, result.status());
        assertEquals("TXN-123", result.transactionId());
    }

    @Test
    void processPayment_returnsFailed_whenAmountIsZero() {
        var result = paymentService.processPayment("ORDER-001", "CUST-001", 0.0);

        assertEquals(PaymentStatus.FAILED, result.status());
        verifyNoInteractions(paymentGateway);
    }

    @Test
    void refundPayment_returnsTrue_whenGatewayConfirmsRefund() {
        when(paymentGateway.refund("TXN-123")).thenReturn(true);

        assertTrue(paymentService.refundPayment("TXN-123"));
    }

    @Test
    void refundPayment_returnsFalse_whenGatewayDeclines() {
        when(paymentGateway.refund("TXN-123")).thenReturn(false);

        assertFalse(paymentService.refundPayment("TXN-123"));
    }
}

File: csharp-workshop/Part4.Behavior.Tests/Service/PaymentServiceTest.cs
using NSubstitute;
using Part4.Behavior.Domain;
using Part4.Behavior.Gateway;
using Part4.Behavior.Service;
using Xunit;

namespace Part4.Behavior.Tests.Service;

public class PaymentServiceTest
{
    private readonly IPaymentGateway _paymentGateway = Substitute.For<IPaymentGateway>();
    private readonly GatewayPaymentService _paymentService;

    public PaymentServiceTest()
    {
        _paymentService = new GatewayPaymentService(_paymentGateway);
    }

    [Fact]
    public void P4_ProcessPayment_ReturnsSuccess_WhenGatewayAcceptsPayment()
    {
        _paymentGateway.Process(Arg.Any<PaymentRequest>())
            .Returns(new PaymentResult("ORDER-001", PaymentStatus.Success, "TXN-123"));

        var result = _paymentService.ProcessPayment("ORDER-001", "CUST-001", 29.0);

        Assert.Equal(PaymentStatus.Success, result.Status);
        Assert.Equal("TXN-123", result.TransactionId);
    }

    [Fact]
    public void P4_ProcessPayment_ReturnsFailed_WhenAmountIsZero()
    {
        var result = _paymentService.ProcessPayment("ORDER-001", "CUST-001", 0.0);

        Assert.Equal(PaymentStatus.Failed, result.Status);
        _paymentGateway.DidNotReceiveWithAnyArgs();
    }

    [Fact]
    public void P4_RefundPayment_ReturnsTrue_WhenGatewayConfirmsRefund()
    {
        _paymentGateway.Refund("TXN-123").Returns(true);

        Assert.True(_paymentService.RefundPayment("TXN-123"));
    }

    [Fact]
    public void P4_RefundPayment_ReturnsFalse_WhenGatewayDeclines()
    {
        _paymentGateway.Refund("TXN-123").Returns(false);

        Assert.False(_paymentService.RefundPayment("TXN-123"));
    }
}
